pip/pip-415.md

```diff
@@ -0,0 +1,112 @@
+- RocketMQ/Kafka: Utilize monotonically increasing numerical offsets for message positioning and acknowledgment
+- Pulsar: Relies on composite MessageID (ledgerId + entryId + batchIndex) for message identification
+
+We propose to add a new API to retrieve the message ID by offset, enabling us to cache the mapping between message ID and offset.
+This will allow us to use offsets for seek and acknowledgment operations when consuming messages through the standardized API.
+
+# Goals
+
+## In Scope
+
+- Add a new API to retrieve the message ID by offset
+
+# High Level Design
+
+# Detailed Design
+The core implementation involves adding a new HTTP endpoint to query MessageId by offset (index). Underlying implementation will leverage existing `Broker Entry Metadata` containing index values.
+
+## Design & Implementation Details
+
+## Public-facing Changes
+
+### Public API
+
+```java
+    @GET
+    @Path("/{tenant}/{namespace}/{topic}/getMessageIdByOffset")
+    @ApiOperation(hidden = true, value = "Get Message ID by offset.",
+            notes = "If the specified offset is a system message, "
+                    + "it will return the message id of the later message.")
+    @ApiResponses(value = {
+            @ApiResponse(code = 307, message = "Current broker doesn't serve the namespace of this topic"),
+            @ApiResponse(code = 403, message = "Don't have admin permission"),
+            @ApiResponse(code = 404, message = "Namespace or partitioned topic does not exist, "
+                    + "or the offset is invalid")})
+    public void getMessageIDByOffsetAndPartitionID(@Suspended final AsyncResponse asyncResponse,
+                                                   @PathParam("tenant") String tenant,
+                                                   @PathParam("namespace") String namespace,
+                                                   @PathParam("topic") @Encoded String encodedTopic,
+                                                   @QueryParam("offset") long offset,
+                                                   @QueryParam("authoritative") @DefaultValue("false")
+                                                   boolean authoritative){
+        validateTopicName(tenant, namespace, encodedTopic);
+        internalGetMessageIDByOffsetAsync(offset, authoritative)
+                .thenAccept(asyncResponse::resume)
+                .exceptionally(ex -> {
+                    if (!isRedirectException(ex)) {
+                        log.error("[{}] Failed to get message id by offset for topic {}, partition id {}, offset {}",
+                                clientAppId(), topicName, offset, ex);
+                    }
+                    resumeAsyncResponseExceptionally(asyncResponse, ex);
+                    return null;
+                });
+    }
+```
+
+### Binary protocol
+
+### Configuration
+
+### CLI
+Add new subcommand to pulsar-admin:
+
+```bash
+    pulsar-admin topics get-message-id-by-offset \
```

Comment:
Since we're using "index" as the concept, it's better to replace "offset" with "index" in all commands and methods.